File: bindings/python/insight/signal/convolution.py
"""Convolution and correlation functions for signal processing.

Provides FFT-accelerated convolution/correlation for 1-D and 2-D signals,
as well as utilities for choosing the best convolution method.
"""

from insight._insight import signal as _signal

__all__ = [
    "fftconvolve",
    "correlate",
    "convolve2d",
    "correlate2d",
    "choose_conv_method",
    "correlation_lags",
]


def fftconvolve(in1, in2, mode="full"):
    """Convolve two N-D arrays using FFT.

    Args:
        in1: First input array.
        in2: Second input array.
        mode: Output mode.  One of ``'full'`` (default), ``'same'``, or
            ``'valid'``.

    Returns:
        Convolution result array.
    """
    return _signal.fftconvolve(in1, in2, mode=mode)


def correlate(in1, in2, mode="full"):
    """Cross-correlate two N-D arrays.

    Args:
        in1: First input array.
        in2: Second input array.
        mode: Output mode.  One of ``'full'`` (default), ``'same'``, or
            ``'valid'``.

    Returns:
        Cross-correlation result array.
    """
    return _signal.correlate(in1, in2, mode=mode)


def convolve2d(in1, in2, mode="full"):
    """Convolve two 2-D arrays.

    Args:
        in1: First 2-D input array.
        in2: Second 2-D input array.
        mode: Output mode.  One of ``'full'`` (default), ``'same'``, or
            ``'valid'``.

    Returns:
        2-D convolution result array.
    """
    return _signal.convolve2d(in1, in2, mode=mode)


def correlate2d(in1, in2, mode="full"):
    """Cross-correlate two 2-D arrays.

    Args:
        in1: First 2-D input array.
        in2: Second 2-D input array.
        mode: Output mode.  One of ``'full'`` (default), ``'same'``, or
            ``'valid'``.

    Returns:
        2-D cross-correlation result array.
    """
    return _signal.correlate2d(in1, in2, mode=mode)


def choose_conv_method(in1, in2, mode="full"):
    """Choose the fastest convolution method for the given inputs.

    Args:
        in1: First input array.
        in2: Second input array.
        mode: Output mode.  One of ``'full'`` (default), ``'same'``, or
            ``'valid'``.

    Returns:
        A string indicating the recommended method (``'fft'`` or
        ``'direct'``).
    """
    return _signal.choose_conv_method(in1, in2, mode=mode)


def correlation_lags(in1_len, in2_len, mode="full"):
    """Compute the lag vector for cross-correlation.

    Args:
        in1_len: Length of the first input array.
        in2_len: Length of the second input array.
        mode: Correlation mode.  One of ``'full'`` (default), ``'same'``, or
            ``'valid'``.

    Returns:
        1-D array of lag indices.
    """
    return _signal.correlation_lags(in1_len, in2_len, mode=mode)

File: bindings/lua/insight/signal/convolution.lua
--- Convolution and correlation functions.
-- Provides FFT-based convolution, cross-correlation,
-- and 2-D convolution/correlation operations.
-- @module insight.signal.convolution

local native = require("_insight")
local sig = native.signal

local M = {}

local function _wrap(names, fn)
  return function(...)
    if select("#", ...) == 1 and type(select(1, ...)) == "table" then
      local t = select(1, ...)
      local has_names = false
      for k, _ in pairs(t) do
        if type(k) ~= "number" then
          has_names = true
          break
        end
      end
      if has_names then
        local pos = {}
        for i, name in ipairs(names) do
          pos[i] = t[name]
          if pos[i] == nil then
            pos[i] = t[i]
          end
        end
        return fn(table.unpack(pos, 1, #names))
      end
    end
    return fn(...)
  end
end

--- Convolve two N-dimensional arrays using FFT.
-- @tparam Array a First input array.
-- @tparam Array b Second input array.
-- @string[opt="full"] mode Convolution mode: 'full', 'valid', or 'same'.
-- @treturn Array Convolved array.
M.fftconvolve = _wrap({ "a", "b", "mode" }, function(a, b, mode)
  return sig.fftconvolve(a, b, mode or "full")
end)

--- Cross-correlate two N-dimensional arrays.
-- @tparam Array a First input array.
-- @tparam Array b Second input array.
-- @string[opt="full"] mode Correlation mode: 'full', 'valid', or 'same'.
-- @treturn Array Cross-correlation result.
M.correlate = _wrap({ "a", "b", "mode" }, function(a, b, mode)
  return sig.correlate(a, b, mode or "full")
end)

--- 2-D convolution of two arrays.
-- @tparam Array a First input 2-D array.
-- @tparam Array b Second input 2-D array.
-- @string[opt="full"] mode Convolution mode: 'full', 'valid', or 'same'.
-- @string[opt="fill"] boundary Boundary handling: 'fill', 'wrap', 'symm'.
-- @number[opt=0] fillvalue Value for fill boundary.
-- @treturn Array 2-D convolved array.
M.convolve2d = _wrap({ "a", "b", "mode", "boundary", "fillvalue" }, function(a, b, mode, boundary, fillvalue)
  return sig.convolve2d(a, b, mode or "full", boundary or "fill", fillvalue or 0)
end)

--- 2-D cross-correlation of two arrays.
-- @tparam Array a First input 2-D array.
-- @tparam Array b Second input 2-D array.
-- @string[opt="full"] mode Correlation mode: 'full', 'valid', or 'same'.
-- @string[opt="fill"] boundary Boundary handling: 'fill', 'wrap', 'symm'.
-- @number[opt=0] fillvalue Value for fill boundary.
-- @treturn Array 2-D cross-correlation result.
M.correlate2d = _wrap({ "a", "b", "mode", "boundary", "fillvalue" }, function(a, b, mode, boundary, fillvalue)
  return sig.correlate2d(a, b, mode or "full", boundary or "fill", fillvalue or 0)
end)

--- Choose the best convolution method (direct vs FFT).
-- @tparam Array a First input array.
-- @tparam Array b Second input array.
-- @string mode Convolution mode.
-- @treturn string Best method: 'direct' or 'fft'.
M.choose_conv_method = _wrap({ "a", "b", "mode" }, function(a, b, mode)
  return sig.choose_conv_method(a, b, mode)
end)

--- Compute the lag values for cross-correlation.
-- @int input_length Length of the first input.
-- @int kernel_length Length of the second input (kernel).
-- @string[opt="full"] mode Correlation mode.
-- @treturn Array Lag values.
M.correlation_lags = _wrap({ "input_length", "kernel_length", "mode" }, function(input_length, kernel_length, mode)
  return sig.correlation_lags(input_length, kernel_length, mode or "full")
end)

return M
